api/src/main/java/org/apache/iceberg/FieldMetrics.java

```diff
@@ -30,15 +28,15 @@
   private final long valueCount;
   private final long nullValueCount;
   private final long nanValueCount;
-  private final ByteBuffer lowerBound;
-  private final ByteBuffer upperBound;
+  private final Object lowerBound;
+  private final Object upperBound;
```

Comment:
Why change this to `Object` rather than `ByteBuffer`? Seems like conversion to ByteBuffer would be cleaner if this was done in each writer because the writer already has its type because it is going to call the right method on the encoder.

Reply:
I think if we convert this to `ByteBuffer` now we may still need to check type when doing truncation (based on metrics mode), and I think string with non-unicode characters will not vend the same result if truncated by `BinaryUtil.truncateBinary`, so we will either convert the byte buffer back to char sequence and use `UnicodeUtil.truncateString` or create a new `BinaryUtil.truncateString`. Whereas if we do conversion later when evaluating metrics, I think the [code needed for the conversion itself](https://github.com/apache/iceberg/pull/1935/files#diff-32411391d7c2962fe8c090ca56f1e9f63cb9bd473f0e55a9a1581863e9542deaR109) isn't that bad since we know the type of the field, and that's the reason for me to do this change. 

But one thing that may worth noting is that for the current approach, in order ensure the `Conversions.toByteBuffer` could work, for certain writers I have to make sure the min/max from the value writers return the type that `Conversions.toByteBuffer` knows how to translate, if the data type in `write` is not of that type (that is, usage of [this method](https://github.com/apache/iceberg/pull/1935/files#diff-6ccdc1daa757d215170dc57a4b3d4f31a62c8b2993ae2289511e97b2df52ef02R641-R652)). I think we still need to maintain a similar function for translation in each value writer if we return bytebuffer for bounds in field metrics, but it will directly translate input data type to byte buffer instead of doing two hops, and that might be easier to understand.

Reply:
I guess I was thinking that truncation would happen when `FieldMetrics` is constructed, in the leaf writers. If that's not the case, then I think it makes sense to do the conversion later.

If the conversion happens later, then I think this class should be parameterized. I never like to have classes that track just `Object`. We should at least guarantee that both lower and upper bounds are the same type, for example.

Reply:
I was mostly following the pattern of ORC and Parquet to evaluate metrics mode when collecting metrics (which has to be since the file formats collects stats themselves), but I think there's nothing prevent us from ingesting metrics mode during value writers creation, it will just make the visitor pattern a little bit more complicated. I'll give it a try, and thanks for bringing up this idea! 

I guess for now I'll revert the change to `FieldMetrics` in this PR and include it in the next one that updates value writers if we need to change it. Hopefully that doesn't add too much to the next PR!